File: 10. Tree/226.invert-binary-tree.py
#
# @lc app=leetcode id=226 lang=python3
#
# [226] Invert Binary Tree
#
# https://leetcode.com/problems/invert-binary-tree/description/
#
# algorithms
# Easy (75.64%)
# Likes:    13173
# Dislikes: 186
# Total Accepted:    1.8M
# Total Submissions: 2.3M
# Testcase Example:  '[4,2,7,1,3,6,9]'
#
# Given the root of a binary tree, invert the tree, and return its root.
#
#
# Example 1:
#
#
# Input: root = [4,2,7,1,3,6,9]
# Output: [4,7,2,9,6,3,1]
#
#
# Example 2:
#
#
# Input: root = [2,1,3]
# Output: [2,3,1]
#
#
# Example 3:
#
#
# Input: root = []
# Output: []
#
#
#
# Constraints:
#
#
# The number of nodes in the tree is in the range [0, 100].
# -100 <= Node.val <= 100
#
#
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        # return self.recursiveDFS(root)
        # return self.iterativeDFS(root)
        return self.BFS(root)

    def recursiveDFS(self, root):
        if not root:
            return
        # root.left, root.right = root.right, root.left
        # self.invertTree(root.right)
        # self.invertTree(root.left)
        root.left, root.right = self.invertTree(root.right), self.invertTree(root.left)
        return root

    def iterativeDFS(self, root):
        if not root:
            return

        stack = [root]
        while stack:
            node = stack.pop()
            node.right, node.left = node.left, node.right
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        return root

    def BFS(self, root):
        if not root:
            return

        queue = deque([root])
        while queue:
            node = queue.popleft()
            node.right, node.left = node.left, node.right
            if node.right:
                queue.append(node.right)
            if node.left:
                queue.append(node.left)
        return root
# @lc code=end

File: 226.invert-binary-tree.go
/*
 * @lc app=leetcode id=226 lang=golang
 *
 * [226] Invert Binary Tree
 *
 * https://leetcode.com/problems/invert-binary-tree/description/
 *
 * algorithms
 * Easy (77.16%)
 * Likes:    14758
 * Dislikes: 246
 * Total Accepted:    2.7M
 * Total Submissions: 3.4M
 * Testcase Example:  '[4,2,7,1,3,6,9]'
 *
 * Given the root of a binary tree, invert the tree, and return its root.
 *
 *
 * Example 1:
 *
 *
 * Input: root = [4,2,7,1,3,6,9]
 * Output: [4,7,2,9,6,3,1]
 *
 *
 * Example 2:
 *
 *
 * Input: root = [2,1,3]
 * Output: [2,3,1]
 *
 *
 * Example 3:
 *
 *
 * Input: root = []
 * Output: []
 *
 *
 *
 * Constraints:
 *
 *
 * The number of nodes in the tree is in the range [0, 100].
 * -100 <= Node.val <= 100
 *
 *
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
package main

func invertTree(root *TreeNode) *TreeNode {
	// return recursiveDFS(root)
	// return iterativeDFS(root)
	return BFS(root)
}

func recursiveDFS(root *TreeNode) *TreeNode {
	if root != nil {
		root.Left, root.Right = root.Right, root.Left
		invertTree(root.Left)
		invertTree(root.Right)
	}
	return root
}

func iterativeDFS(root *TreeNode) *TreeNode {
	var stack []*TreeNode
	head := root
	for root != nil || len(stack) != 0 {
		for root != nil {
			root.Left, root.Right = root.Right, root.Left
			stack = append(stack, root)
			root = root.Left
		}
		root = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		root = root.Right
	}
	return head
}

func BFS(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	q := []*TreeNode{root}
	for len(q) != 0 {
		node := q[0]
		q = q[1:]
		node.Right, node.Left = node.Left, node.Right
		if node.Right != nil {
			q = append(q, node.Right)
		}
		if node.Left != nil {
			q = append(q, node.Left)
		}
	}
	return root
}

// @lc code=end
